### DISCUSSION TITLE
PASSWORD_RECOVERY never fired

### DISCUSSION THREAD

**Discussion by barrykooij** [supabase/supabase#15951] [Questions]
After I click the link in the password reset my app doesn't have the PASSWORD_RECOVERY event fired (no events are fired at all).

I've created a custom auth provider that has the following code:

```js
    useEffect(() => {
        setLoading(true);
        const getUser = async () => {
            const { data } = await supabase.auth.getUser();
            const { user: currentUser } = data;
            setUser(currentUser ?? null);
            setLoading(false);
        };
        getUser().catch(console.error);


        const { data } = supabase.auth.onAuthStateChange((event: AuthChangeEvent, session: Session | null) => {

            console.log(event);

            switch (event) {
                case 'PASSWORD_RECOVERY':
                    setAuth(false);
                    break;
                case 'SIGNED_IN':
                    // case 'INITIAL_SESSION':
                    if (session !== null) {
                        setUser(session.user);
                        setAuth(true);
                    }

                    break;
                case 'SIGNED_OUT':
                    setUser(null);
                    setAuth(false);
                    break;
            }

        });
        return () => {
            data.subscription.unsubscribe();
        };
    }, []);
```
  

The `supabase.auth.getUser()` crashes / doesn't work when the access_token is in the URL. If I create the client with `detectSessionInUrl: false`, the app does load because `getUser()` doesn't use the URL values. Still no events fired.

What am I missing? Any help is very appreciated.

**Comment by silentworks** [comment]
Are you using this in a NextJS app or a normal React app with react-router? Also note the recommended way for Password reset unless its a SPA without a router is in this guide https://supabase.com/docs/guides/auth/auth-password-reset

**Reply by barrykooij** [reply]
Normal React, it's an SPA with reacter-router-dom. The `createClient()` should resolve the tokens from the URL automatically, right? 

The URL is like this:
`http://localhost:5173/update-password#access_token=TOKEN&expires_in=3600&refresh_token=RTOKEN&token_type=bearer&type=recovery`

I've replaced the access_token en refresh_token for security reasons.

**Reply by barrykooij** [reply]
To make sure there are no other conflicts I've replaced the whole app with just the following code, still no events when the access_token is in the URL.

```js
const Test = () => {
    useEffect(() => {

        const getUser = async () => {
            const { data } = await supabase.auth.getUser();
            console.log("getUser", data);
        };
        getUser();
        
        const { data } = supabase.auth.onAuthStateChange((event: AuthChangeEvent, session: Session | null) => {
            console.log(event);
        });
        console.log(data)
        return () => {
            data.subscription.unsubscribe();
        };
    }, []);

    return (
        <div>test</div>
    );
};

ReactDOM.createRoot(document.getElementById('root') as HTMLElement).render(
    <React.StrictMode>
        <Suspense>
            <Test />
        </Suspense>
    </React.StrictMode>
);
```

**Reply by barrykooij** [reply]
More additional information, while waiting for a reply on this topic I worked on implementing social logins.

These access_tokens are also not being read / used. I feel like the issue is somewhere

**Comment by barrykooij** [comment]
After a deeper dive into supabase and gotrue source then I wanted to do, I've found out the following.

It seems the GoTrueClient is never initialized in `__loadSession`. Or the function is called incorrectly for some reason. The whole process is stuck at ` await this.initializePromise;`

Location:
https://github.com/supabase/gotrue-js/blob/master/src/GoTrueClient.ts#L932

If I remove this line (not a permanent solution of course), Supabase client reads tokens from URL and all works well...

**Reply by barrykooij** [reply]
I'm going to see if I can test the current gotrue-js files because it seems like there are commits that address issues that look a lot like my issue. https://github.com/supabase/gotrue-js/commit/5d142facc8ea60f1b61a6cf6615d8a1b9ba81d0f

**Reply by silentworks** [reply]
I'm going to test this out locally too to see if I run into this issue.

**Reply by barrykooij** [reply]
Didn't solve the issue for me. 

I'm sure the issue for me is the ```await this.initializePromise;```  here https://github.com/supabase/gotrue-js/blob/master/src/GoTrueClient.ts#L932

For some reason this await never "finishes", causing my `__loadSession` to forever wait.

**Reply by barrykooij** [reply]
@silentworks I'd love to hear if you can get this to work in a default React app. My setup is as bareboned as can be:


```js
const supabase = createClient("url", "key");
const Barry = () => {

    useEffect(() => {


        supabase.auth.onAuthStateChange((event, session) => {
            console.log(event);
        });
        console.log("-------------------------");
        console.log("once")

        const t = async () => {
            console.log("START ASYNC");
            const resp = await supabase.auth.getSession();
            console.log("getSession", resp);
            console.log("END ASYNC");
        };
        t();
    }, []);

    return (
        <div>Test 123</div>
    );
};


ReactDOM.createRoot(document.getElementById('root') as HTMLElement).render(
    <Barry />
);
```

For me this never gets passed the `const resp = await supabase.auth.getSession();`.

**Reply by barrykooij** [reply]
After the latest gotruejs version ([2.45.0](https://github.com/supabase/gotrue-js/releases/tag/v2.45.0)), it now works in Firefox and Safari (on Mac). Still doesn't work on Chrome.

The issue is that the `__loadSession` returns `false` on `isInStackGuard('_initialize')`, even when it appears to be in the stackguard. I've logged creating and checking stackguards to the console and  this is the result:

```
GoTrueClient@0 2023-07-23T09:40:57.674Z #_acquireLock begin -1
helpers.ts:365 CREATE stackGuard: ENV_CHECK
helpers.ts:340 FOUND stackGuard: ENV_CHECK
helpers.ts:345 NOT FOUND stackGuard: _acquireLock
GoTrueClient.ts:236 GoTrueClient@0 2023-07-23T09:40:57.684Z #_acquireLock lock acquired for storage key sb-hidden-auth-token
helpers.ts:365 CREATE stackGuard: _acquireLock
helpers.ts:365 CREATE stackGuard: _initialize
GoTrueClient.ts:236 GoTrueClient@0 2023-07-23T09:40:57.684Z #_initialize() begin is PKCE flow false
GoTrueClient.ts:236 GoTrueClient@0 2023-07-23T09:40:57.684Z Getting session from URL.
GoTrueClient.ts:236 GoTrueClient@0 2023-07-23T09:40:57.684Z #_useSession begin
helpers.ts:345 NOT FOUND stackGuard: _useSession
GoTrueClient.ts:236 GoTrueClient@0 2023-07-23T09:40:57.684Z #_acquireLock begin -1
helpers.ts:345 NOT FOUND stackGuard: _acquireLock
GoTrueClient.ts:236 GoTrueClient@0 2023-07-23T09:40:57.684Z #_acquireLock lock acquired for storage key sb-hidden-auth-token
helpers.ts:365 CREATE stackGuard: _acquireLock
helpers.ts:365 CREATE stackGuard: _useSession
GoTrueClient.ts:236 GoTrueClient@0 2023-07-23T09:40:57.684Z #__loadSession() begin
GoTrueClient.ts:236 GoTrueClient@0 2023-07-23T09:40:57.684Z stackGuardsSupported() true
helpers.ts:345 NOT FOUND stackGuard: _initialize
```

I'll create an issue on the gotruejs repository.

**Reply by barrykooij** [reply]
Issue created: https://github.com/supabase/gotrue-js/issues/744

**Reply by silentworks** [reply]
Yeah I'm getting the same behavior as your app. The PASSWORD_RECOVERY event is not getting fired at all.

**Reply by Yermanusser** [reply]
> I'm going to test this out locally too to see if I run into this issue.
